### PR TITLE
Add label tags for authenticators.

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
**Description:**
Add a tag list for authenticators and IDPs. The tags would indicate the characteristics of the each authenticator/IDP.

Related to #https://github.com/wso2/product-is/issues/11948

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java [hunk 1]
```diff
@@ -386,4 +386,17 @@ public String getAuthMechanism() {
         diagnosticLog.info("AuthMechanism is set to : " + authMechanism + " for the authenticator: " + getName());
         return authMechanism;
     }
+
+    @Override
+    public String[] getTags() {
+
+        String tags = getAuthenticatorConfig().getParameterMap().get(FrameworkConstants.TAGS);
+        String[] tagList;
+        if (StringUtils.isEmpty(tags)) {
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java:395]: we can use StringUtils.split which handles the empty case

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java [hunk 2]
```diff
@@ -40,7 +40,7 @@
      * @param request
      * @return boolean
      */
-    public boolean canHandle(HttpServletRequest request);
```
Comment (emswbandara) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java:43]: what is the reason for this change?

File: features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json [hunk 3]
```diff
@@ -66,10 +66,12 @@
   "authentication.authenticator.basic.name": "BasicAuthenticator",
   "authentication.authenticator.basic.enable": true,
   "authentication.authenticator.basic.parameters.AuthMechanism": "basic",
+  "authentication.authenticator.basic.parameters.Tags": "Username-Password",
 
   "authentication.authenticator.basic_request_path.name": "BasicAuthRequestPathAuthenticator",
   "authentication.authenticator.basic_request_path.enable": true,
   "authentication.authenticator.basic_request_path.parameters.AuthMechanism": "basic",
+  "authentication.authenticator.basic_request_path.parameters.Tags": "Request path",
```
Comment (emswbandara) [inline features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json:74]: better to use a unified format for tags. if we follow the pattern in "Username-Password" to use '-' as delimiter and start each word with a capital letter, then better to use that pattern in all the places

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java (hunk 2 above)
Comment (dewniMW) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java:43]: since this is an interface, defining it as public is redundant.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java [hunk 4]
```diff
@@ -98,4 +98,14 @@ public AuthenticatorFlowStatus process(HttpServletRequest request,
     default String getAuthMechanism() {
         return getName();
     }
+
+    /**
+     * Get the tags of the authenticators.
+     *
+     * @return List of Tags
```
Comment (dewniMW) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java:105]:
```suggestion
     * @return List of tags.
```

File: features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json [hunk 5]
```diff
@@ -81,43 +83,51 @@
...
   "authentication.authenticator.google.parameters.GoogleAuthzEndpoint": "https://accounts.google.com/o/oauth2/auth",
   "authentication.authenticator.google.parameters.GoogleUserInfoEndpoint": "https://www.googleapis.com/oauth2/v3/userinfo",
+  "authentication.authenticator.google.parameters.Tags": "Social Login",
 
   "authentication.authenticator.microsoft_live.name": "MicrosoftWindowsLive",
   "authentication.authenticator.microsoft_live.enable": true,
   "authentication.authenticator.microsoft_live.parameters.AuthTokenEndpoint": "https://login.live.com/oauth20_token.srf",
   "authentication.authenticator.microsoft_live.parameters.AuthnEndpoint": "https://login.live.com/oauth20_authorize.srf",
   "authentication.authenticator.microsoft_live.parameters.UserInfoEndpoint": "https://apis.live.net/v5.0/me?access_token=",
+  "authentication.authenticator.microsoft_live.parameters.Tags": "Social Login,OIDC",
```
Comment (dewniMW) [inline features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json:110]: is OIDC required here? asking this based on the tags defined for google

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java [hunk 6]
```diff
@@ -202,4 +214,20 @@ public int hashCode() {
     public boolean isValidPropertyValue(Property property) {
         return property != null && StringUtils.isNotBlank(property.getValue());
     }
+
+    /**
+     * @return
```
Comment (ashensw) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java:219]: Let's add the method comments properly.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java [hunk 7]
```diff
@@ -194,4 +207,20 @@ public boolean equals(Object o) {
     public int hashCode() {
         return name != null ? name.hashCode() : 0;
     }
+
+    /**
+     * @return tags
```
Comment (ashensw) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java:212]: Let's add the method comments properly. Check all the new methods.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java (hunk 1 above)
Comment (Thisara-Welmilla) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java]: Addressed with commit ad110cdff4ffcfe5766b31f1a09bebaf633730e6

File: features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json (hunk 3 above)
Comment (Thisara-Welmilla) [reply features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json]: Addressed with commit ad110cdff4ffcfe5766b31f1a09bebaf633730e6

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/ApplicationAuthenticator.java]: Addressed with commit ad110cdff4ffcfe5766b31f1a09bebaf633730e6

File: features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json (hunk 5 above)
Comment (Thisara-Welmilla) [reply features/authentication-framework/org.wso2.carbon.identity.application.authentication.framework.server.feature/resources/org.wso2.carbon.identity.application.authentication.framework.server.feature.default.json]: Addressed with commit ad110cdff4ffcfe5766b31f1a09bebaf633730e6

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java (hunk 6 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java]: Addressed with commit ad110cdff4ffcfe5766b31f1a09bebaf633730e6

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java (hunk 7 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java]: Addressed with commit ad110cdff4ffcfe5766b31f1a09bebaf633730e6

**Review by dewniMW — APPROVED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java [hunk 8]
```diff
@@ -82,6 +86,9 @@ public static LocalAuthenticatorConfig build(OMElement localAuthenticatorConfigO
                 if (member.getText() != null && member.getText().trim().length() > 0) {
                     localAuthenticatorConfig.setEnabled(Boolean.parseBoolean(member.getText()));
                 }
+            } else if ("Tags".equals(member.getLocalName())) {
```
Comment (DinikaSen) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java:89]: Can't we use the TAG constant that is introduced https://github.com/wso2/carbon-identity-framework/pull/3570/files#diff-9461270f09d52c105a65affa1fdc9fbd740274afc25204ffd5ec2bf49e6975b9R133 in this place instead of the string "Tags"?

Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java:89]: Address with https://github.com/wso2/carbon-identity-framework/pull/3779 PR
